src/Cli/dotnet/commands/dotnet-sln/add/Program.cs

```diff
@@ -2,143 +2,183 @@
+            SolutionModel solution = await serializer.OpenAsync(solutionFileFullPath, cancellationToken);
+            // set UTF8 BOM encoding for .sln
+            if (serializer is ISolutionSerializer<SlnV12SerializerSettings> v12Serializer)
             {
-                projectFilePath = projectFilePath.Substring(currentDirString.Length);
+                solution.SerializerExtension = v12Serializer.CreateModelExtension(new()
+                {
+                    Encoding = new UTF8Encoding(encoderShouldEmitUTF8Identifier: true)
+                });
             }
-
-            var projectDirectoryPath = TrimProject(projectFilePath);
-            if (string.IsNullOrEmpty(projectDirectoryPath))
-                return solutionFolders;
-
-            var solutionFoldersPath = TrimProjectDirectory(projectDirectoryPath);
-            if (string.IsNullOrEmpty(solutionFoldersPath))
-                return solutionFolders;
-
-            solutionFolders.AddRange(solutionFoldersPath.Split(Path.DirectorySeparatorChar));
-
-            return solutionFolders;
-        }
-
-        private IList<string> DetermineSolutionFolder(SlnFile slnFile, string fullProjectPath)
-        {
-            if (_inRoot)
+            // Set default configurations and platforms for sln file
+            foreach (var platform in new[]{ "Any CPU", "x64", "x86" })
             {
-                // The user requested all projects go to the root folder
-                return null;
+                solution.AddPlatform(platform);
             }
-
-            if (_relativeRootSolutionFolders != null)
+            foreach (var buildType in new []{ "Debug", "Release" })
             {
-                // The user has specified an explicit root
-                return _relativeRootSolutionFolders;
+                solution.AddBuildType(buildType);
             }
 
-            // We determine the root for each individual project
-            var relativeProjectPath = Path.GetRelativePath(
-                PathUtility.EnsureTrailingSlash(slnFile.BaseDirectory),
-                fullProjectPath);
+            SolutionFolderModel? solutionFolder = (!_inRoot && _solutionFolderPath != null)
+                ? solution.AddFolder(GetSolutionFolderPathWithForwardSlashes(_solutionFolderPath))
+                : null;
 
-            return GetSolutionFoldersFromProjectPath(relativeProjectPath);
+            foreach (var projectPath in projectPaths)
+            {
+                string relativePath = Path.GetRelativePath(Path.GetDirectoryName(solutionFileFullPath), projectPath);
+                try
+                {
+                    AddProject(solution, relativePath, projectPath, solutionFolder);
+                }
+                catch (InvalidProjectFileException ex)
+                {
+                    Reporter.Error.WriteLine(string.Format(CommonLocalizableStrings.InvalidProjectWithExceptionMessage, projectPath, ex.Message));
```

Comment:
These are nonfatal? They're nonfatal because another project might have been successfully added?

Reply:
These are non-fatal as to comply with current behavior and because there are tests expecting this. But we could further discuss this if needed:)